pkg/mmesh/etcd_secret.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 IBM Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package mmesh
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	EtcdConnectionKey = "etcd_connection"
+)
+
+// An EtcdSecret represents the etcd configuation for a user namespace
+type EtcdSecret struct {
+	Log                 logr.Logger
+	Name                string
+	Namespace           string
+	ControllerNamespace string
+	EtcdConfig          *EtcdConfig
+	Scheme              *runtime.Scheme
+}
+
+func (es EtcdSecret) Apply(ctx context.Context, cl client.Client) error {
+	commonLabelValue := "modelmesh-controller"
+
+	s := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      es.Name,
+			Namespace: es.Namespace,
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": commonLabelValue,
+			},
+		},
+	}
+	es.addData(s)
+	err := cl.Create(ctx, s)
+	if err != nil && errors.IsAlreadyExists(err) {
+		err = cl.Update(ctx, s)
+	}
+
+	return err
+}
+
+// Add data to the provided secret
+func (es EtcdSecret) addData(s *corev1.Secret) {
+	etcdEndpoints := strings.Split(es.EtcdConfig.Endpoints, ",")
+	newEtcdEndpoints := ""
+
+	// For each endpoint, insert controller namespace if not there
+	for i := range etcdEndpoints {
+		re := regexp.MustCompile(`(?:https?://)?([^\\s.:]+)(:\\d+)?`)
+		if se := re.FindStringSubmatchIndex(etcdEndpoints[i]); se != nil {
+			etcdEndpoints[i] = fmt.Sprintf("%s.%s%s", etcdEndpoints[i][:se[3]], es.ControllerNamespace, etcdEndpoints[i][se[3]:])
+		}
+		if i != 0 {
+			newEtcdEndpoints += ","
+		}
+		newEtcdEndpoints += etcdEndpoints[i]
+	}
+
+	es.EtcdConfig.Endpoints = newEtcdEndpoints
+	es.EtcdConfig.RootPrefix = fmt.Sprintf("%s/mm_ns/%s", es.EtcdConfig.RootPrefix, es.Namespace)
+
+	b, _ := json.Marshal(es.EtcdConfig)
```

Comment:
Sorry forgot to add this in last review - should check for the error here and propagate it back